DOC Add tutorial for adding a new package (#1829)
pyodide/pyodide: docs/development/meta-yaml.md

=(meta-yaml-spec)

# The meta.yaml specification

Packages are defined by writing a `meta.yaml` file. The format of these files is
based on the `meta.yaml` files used to build [Conda
packages](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html),
though it is much more limited. The most important limitation is that Pyodide
assumes there will only be one version of a given library available, whereas
Conda allows the user to specify the versions of each package that they want to

> install. Despite the limitations, keeping the file format as close as possible
> to conda's should make it easier to use existing conda package definitions as a
> starting point to create Pyodide packages. In general, however, one should not

Reviewer: ```suggestion
install. Despite the limitations, it is recommended to use existing conda 
package definitions as a starting point to create Pyodide packages. In
general, however, one should not
```

> expect Conda packages to "just work" with Pyodide, see `pr`{795}`

Reviewer: ```suggestion
expect Conda packages to "just work" with Pyodide, see {pr}`795`
```

The supported keys in the `meta.yaml` file are described below.

## `package`

### `package/name`

The name of the package. It must match the name of the package used when
expanding the tarball, which is sometimes different from the name of the package
in the Python namespace when installed. It must also match the name of the
directory in which the `meta.yaml` file is placed. It can only contain

> alpha-numeric characters and `-`, `_`.

Reviewer: ```suggestion
alpha-numeric characters, `-`, and `_`.
```

At most one checksum entry should be provided per package.

### `source/patches`

A list of patch files to apply after expanding the tarball. These are applied
using `patch -p1` from the root of the source tree.

### `source/extras`

Extra files to add to the source tree. This should be a list where each entry is
a pair of the form `(src, dst)`. The `src` path is relative to the directory in
which the `meta.yaml` file resides. The `dst` path is relative to the root of
source tree (the expanded tarball).

## `build`

### `build/skip_host`

Skip building C extensions for the host environment. Default: `True`.

Setting this to `False` will result in ~2x slower builds for packages that
include C extensions. It should only be needed when a package is a build
time dependency for other packages. For instance, numpy is imported during
installation of matplotlib, importing numpy also imports included C extensions,
therefore it is built both for host and target.

### `build/cflags`

Extra arguments to pass to the compiler when building for WebAssembly.

(This key is not in the Conda spec).

### `build/cxxflags`

> Extra arguments to pass to the compiler when building C++ files for WebAssembly. Note that both clfags and cxxflags will be used when compiling C++ files. A common example would be to use -std=c++11 for code that makes use of C++11 features.

Reviewer: ```suggestion
Extra arguments to pass to the compiler when building C++ files for WebAssembly.
Note that both `cflags` and `cxxflags` will be used when compiling C++ files. 
A common example would be to use `-std=c++11` for code that makes use of C++11 features.
```
